lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,464 @@
+
+--------------------------------------------------------------------------------
+-- Constructors
+--------------------------------------------------------------------------------
+
+-- | Construct the initial delegation state.
+initialDelegationState
+    :: k 'AccountK XPub
+    -> DelegationState k
+initialDelegationState accK = DelegationState accK Zero
+
+--------------------------------------------------------------------------------
+-- Chain
+--------------------------------------------------------------------------------
+
+-- | A transaction type specific to `DelegationState`.
+--
+-- Intented to be converted both from and to a more real transaction type.
+--
+-- When constructing a real transaction from `Tx`, these `outputs`
+-- should appear before other outputs. In the theoretical event that there's
+-- also a user-specified output with the same payment key as the pointer output,
+-- `applyTx` will track the first one as the new pointer.
+data Tx = Tx
+    { certs :: [Cert]
+    , inputs :: [(TxIn, Coin)]
+    , outputs :: [TxOut]
+    }
+    deriving (Eq, Generic)
+    deriving Show via (Quiet Tx)
+
+instance Semigroup Tx where
+    (Tx cs1 is1 os1) <> (Tx cs2 is2 os2) =
+        Tx (cs1 <> cs2) (is1 <> is2) (os1 <> os2)
+
+data Cert
+    = RegisterKey RewardAccount
+    | Delegate RewardAccount
+      -- ^ Which pool we're delegating to is here (and for now) irrelevant.
+      -- The main thing is that there exists a witness on-chain for this stake
+      -- key (registration certs don't require witnesses)
+      --
+      -- TODO: We may also want to add the PoolId here.
+    | DeRegisterKey RewardAccount
+    deriving (Eq, Show, Generic)
+
+-- | Given a `DelegationState`, produce a `Tx` registering or de-registering
+-- stake-keys, in order to have @n@ stake-keys.
+--
+-- E.g. @setPortfolioOf s0 _ _ 2@ creates a tx which after application causes
+-- the state to have @activeKeys == [0,1]@
+--
+-- Returns @Nothing@ if the target @n@ is already reached.
+setPortfolioOf
+    :: (SoftDerivation k, ToRewardAccount k)
+    => DelegationState k
+    -> Coin
+        -- ^ minUTxOVal
+    -> (k 'AddressK XPub -> Address)
+        -- ^ A way to construct an Address
+    -> (RewardAccount -> Bool)
+        -- ^ Whether or not the key is registered.
+        --
+        -- TODO: Need a Set or Map for the real implementation with LSQ.
+    -> Int
+        -- ^ Target number of stake keys.
+    -> Maybe Tx
+setPortfolioOf s minUTxOVal mkAddress isReg n =
+    repairKey0IfNeededTx <> changeStateTx
+  where
+    repairKey0IfNeededTx :: Maybe Tx
+    repairKey0IfNeededTx = case repairKey0 $ state s of
+        [] -> Nothing
+        cs -> Just $ Tx cs [] []
+      where
+        repairKey0 (More _ _ False) | n > 0 = deleg [minBound]
+        repairKey0 _ = []
+
+    changeStateTx :: Maybe Tx
+    changeStateTx = txWithCerts $ case compare (toEnum n) (nextKeyIx s) of
+        GT -> deleg [nextKeyIx s .. toEnum (n - 1)]
```

Comment:
Took me some time to convince myself that the next index was inded `toEnum (n - 1)` partly because it feels to me that the logic of `nextKeyIx` is sort of "leaking out". That's the kind of moment where I think functions like `scanl / iterate` fit nicely to avoid silly mistake, imagine a function `nextState :: DelegationState -> DelegationState`, such that it can then be written as:

```hs
take n $ iterate nextState s 
```

This reads much better to me as in "give me the next `n` states"; and the same approach can be done when going in the other direction using a `predState` analogous function.

Reply:
I tried adding
```haskell
nextState :: (Index 'Soft 'AddressK -> pointer) -> State pointer -> State pointer
prevState :: (Index 'Soft 'AddressK -> pointer) -> State pointer -> State pointer
```
to be able to advance the `State` with a `TxOut` as a pointer, rather than a `TxIn`, and also re-use them in `applyTx`.

But there's also the problem of:
1. Do we call `nextState` or `prevState`, and when do we stop?
2. How to yield certificates, which might need to be modified based on `isReg`.

It seemed like a `iterate`-approach would end up very complicated, potentially with many `foldl`, `scanl` and `take` steps. And boundary-cases of `takeWhile` seems just as complicated as `.. toEnum (n - 1)`.

So without an "aha" idea… I think being somewhat imperative here is fine… The tests should cover a lot of things, and give us confidence.